src/figure_finder.py

```diff
@@ -0,0 +1,185 @@
+            plane = pyrr.plane.create()
+            intersection_point = pyrr.geometric_tests.ray_intersect_plane(ray_local_frame, plane)
+            if intersection_point is not None:
+                return intersection_point
+            rospy.logwarn("The line is parallel to the plane, cannot get intersection.")
+
+        return (0, 0, 0)
+
+    def create_figures(self, frame, bboxes: List[BoundingBoxLabeled], config):
+        def shrink_bbox(bbox_labeled: BoundingBoxLabeled, offset_percent: float = 0.0):
+            bbox_labeled.bbox.size_x = int(offset_percent * (bbox_labeled.bbox.size_x))
+            bbox_labeled.bbox.size_y = int(offset_percent * (bbox_labeled.bbox.size_y))
+
+        def get_img_piece(frame, bbox_labeled: BoundingBoxLabeled):
+            x1 = int(bbox_labeled.bbox.center.x - bbox_labeled.bbox.size_x / 2)
+            x2 = int(bbox_labeled.bbox.center.x + bbox_labeled.bbox.size_x / 2)
+            y1 = int(bbox_labeled.bbox.center.y - bbox_labeled.bbox.size_y / 2)
+            y2 = int(bbox_labeled.bbox.center.y + bbox_labeled.bbox.size_y / 2)
+            return frame[y1:y2, x1:x2]
+
+        figures = []
+
+        for bbox_labeled in bboxes:
+            try:
+                # Filter by bbox width height ratio
+                if bbox_labeled.label in config["ratio_verification"]["labels"]:
+                    ratio = bbox_labeled.bbox.size_x / bbox_labeled.bbox.size_x
+                    if ratio < 0.8 or ratio > 1.25:
+                        print(f"Ratio excedeed: {ratio}")
+                        continue
+
+                # Filter by bbox area to frame ratio
+                if bbox_labeled.label in config["area_verification"]["labels"]:
+                    frame_area = frame.shape[0] * frame.shape[1]
+                    bbox_to_image_ratio = bbox_labeled.bbox.size_x * bbox_labeled.bbox.size_y / frame_area
+                    if bbox_to_image_ratio < config["area_verification"]["min_bbox_to_image_ratio"]:
+                        print(f"Bbox area to frame ratio too small: {bbox_to_image_ratio}")
+                        continue
+
+                color = None
+                determined_type = None
+
+                if bbox_labeled.label in config["bbox_shrink"]["labels"]:
+                    shrink_bbox(bbox_labeled, 0.5)
+
+                if bbox_labeled.label in config["ball_color_verification"]["labels"]:
+                    figure_img = get_img_piece(frame, bbox_labeled)
+                    dominant_colors = self.color_detection.get_dominant_colors(figure_img, 2, show=False)
+                    color = self.color_detection.get_matching_color(dominant_colors)
+                    determined_type = config["ball_color_verification"]["color_mapping"].get(color, "unknown")
+
+                if bbox_labeled.label in config["label_mapping"]["labels"]:
+                    determined_type = config["label_mapping"]["mapping"].get(bbox_labeled.label, "unknown")
```

Comment:
Can you move all the verification steps to a separate function to build a nice pipeline?

Reply:
It will be done during making a visualizer of verification